core/string/ustring.h

```diff
@@ -699,6 +699,188 @@ _FORCE_INLINE_ int64_t str_compare(const L *l_ptr, const R *r_ptr) {
 	}
 }
 
+// Specifies how a given value should be string-formatted. See FormatSpec documentation for context
+// and semantics.
+struct FormatSpecPart {
+	// Minimum field width, including prefixes, separators, and other formatting characters.
+	int width = 0;
+
+	enum AlignType {
+		ALIGN_DEFAULT, // Default per formatted value type. Right for numbers, left for other types.
+		ALIGN_LEFT, // '<'
+		ALIGN_RIGHT, // '>'
+		ALIGN_CENTER, // '^'
+		ALIGN_PAD_AFTER_SIGN, // '='. For numbers, pad after the sign and before the number.
+	};
+	AlignType align_type = ALIGN_DEFAULT;
+	String align_pad = " ";
+
+	enum SignType {
+		SIGN_NEGATIVE_ONLY, // '-'. Show sign only for negative numbers. Default.
+		SIGN_ALWAYS, // '+'. Always show sign.
+		SIGN_SPACE_FOR_PLUS, // ' '. Show space for positive numbers, sign for negative numbers.
+	};
+	SignType sign_type = SIGN_NEGATIVE_ONLY;
+
+	// Add '0b', '0o', '0x', '0X' prefixes for binary, octal, hexadecimal types.
+	bool alternate_form = false; // '#'
+
+	// Separate floats and decimals by groups of 3 digits, binary, octal, hex by 4 digits.
+	// group_separator == "" means no grouping.
+	String group_separator = "";
+
+	// For FLOAT type, precision is the number of digits to be displayed _after_ the decimal point. Default 6.
+	// For FLOAT_SIG_FIG type, precision is the number of digits before _and_ after the decimal point.
+	// For STRING type, precision is the maximum field size.
+	// For int types, precision is not allowed.
+	// Negative precision indicates precision is not specified (use defaults).
+	int precision = -1;
```

Comment:
```suggestion
	uint8_t precision = 0;
```

I think using 0 to denote infinite precision is fine? Otherwise we can set it to 255 by default.

Reply:
0 is valid distinct from 'default'. Made this std::optional<uint8_t> to handle default case.